src/queue/immediate.ts

```diff
@@ -87,6 +87,134 @@ export default class Immediate extends Queue {
     return null
   }
 
+  public bookThreadByIndexes(indexes: number[]) {
+    for (const item of indexes) {
+      if (this.threads[item] === null) {
+        this.logger.silly('booked thread ' + item)
+        this.threads[item] = 'booked'
+        return item;
+      }
+    }
+    return null
+  }
+
+  public getThreadsByCriteriaTags(threadeIndex, configuration, tags: string[], host) {
+    let include = false
+    let exclude = false
+    // testujeme stitky pokud jsou nastaveny
+    if (tags && tags.length > 0) {
+      for (const item in tags) {
+        if (configuration[threadeIndex].include.includes(tags[item])) {
+          include = true
+        }
+        if (configuration[threadeIndex].exclude.includes(tags[item])) {
+          exclude = true
+          break
+        }
+      }
+    }
+
+    if ((include === true && exclude === false) ||
+      ((tags === undefined || tags.length === 0) && configuration[threadeIndex].include.length === 0) ||
+      (configuration[threadeIndex].include.length === 0 && exclude === false &&
+        (configuration[threadeIndex].implementation === null || configuration[threadeIndex].implementation === host))
+    ) {
+      // vlozeno pres shodu v include a zaroven neni vylouceno pres exclude
+      // pokud proces nema stitek a vlakno nema nastaveno co do nej muze pres include
+      // pokud vlakno nedefinuje include a proces nevypadne pres exclude a shoduje se host nebo neni definovan
+      return threadeIndex
+    }
+
+    return null
+  }
+
+  public getThreadsByCriteria(host, tags: string[]) {
+    const threads = []
+    const configuration = this.nconf.get('immediate:threads')
+    for (let i = 0; i < configuration.length; i++) {
+      const threadeIndex = this.getThreadsByCriteriaTags(i, configuration, tags, host)
+      if (threadeIndex !== null) {
+        threads.push(threadeIndex);
+      }
+    }
```

Comment:
Tady by se mi vic libilo udelat to jako immutable pipelinu, viz např. https://martinfowler.com/articles/refactoring-pipelines.html - tzn. nejspíš dvojici map/filter v tomhle případě. Ale funguje to stejně a je to jen otázka čitelnosti a nejěké konvence v dané codebase.